### ISSUE TITLE
Data preprocessing

### ISSUE THREAD

**Issue by adelbibi** [CVRL/iris-recognition-OTS-DNN#1]
Hello, 

Thanks for your work.

The model you provide for DRN assumes the input to the network is 3 dimensional, e.g. RBB data. This is consistent with the architecture provided in DRN. However, the majority of iris images/data trained on, e.g. CASIA, are gray scale images.

How do you transform/pre-process the gray scale data to 3 dimensional input?

Best,

**Comment by DanielKerrigan** [comment]
Hi,

If I remember correctly, we copied the grayscale image into the R, G, and B channels so that all 3 channels had the same values.

I will try to double check that this is the approach that we took and I will let you know if we used a different process.

Sorry for not explaining that in the paper.

Best,
Dan

**Comment by adelbibi** [comment]
Thank you for the effort, Daniel.

I believe there must be some sort of normalization you have done, at least during test time. For instance, after loading the DRN parameters you provided to DRN model, I tried evaluating the model on some of the images used in training (CASIA-V4) to assure the model loading is run correctly.

Here is an example of the results I get.
https://drive.google.com/drive/folders/1zd8mg75_erLUfrX_jQkJmj0IoO_mZ_0G?usp=sharing


test_sample.png: An image sample from the CASIA dataset.
class_0.png: The prediction network output image for class 0.
class_1.png: The prediction network output image for class 1.
class0_bigger_than_class1.png: The maximum among those two channels.


Note that the maximum among those two classes is always 0; indicating that one of the two classes is always higher than the other class. There are of course variations per class (as seen in class0.png and clas1.png); however, the network is still confident that one class is always the prediction over the other class.

I believe this is an indication of some form of input normalization I'm not aware of. The current normalization I perform is simply that I replicate the gray scale image to 3 channels and then assure that the input range is 0-1 double precision.

For example, could it be that you also subtract from all input data during test time some (mean representing the average image in training data) and scaling by training data standard deviation?

Your input on this is highly appreciated.

**Comment by DanielKerrigan** [comment]
Hello,

I believe that you are correct. The [DRN readme](https://github.com/fyu/drn#prepare-data) that we link to describes an `info.json` file that contains the mean and standard deviation of the training data.

I forgot that this file was used during testing. I was thinking that it was only used during training. My apologies.

I fear that I may not have saved the `info.json` file that we used. My sincere apologies for this. After graduating last year, I lost access to the computers the training was performed on and it appears that I didn't copy that file over.

I do have all of the training data, however, so I should be able to recreate the `info.json` if you need it.

Sorry for the troubles,
Dan

**Comment by adelbibi** [comment]
Thanks for your response, Dan.

First, congrats on the graduation Dr Daniel. ;)

As for the info file, it would be of a great help if you can provide that file only if that does not take much of your time. I believe it will also be helpful for other people in the future. 

On a side note, I have been training the model again with the checkpoint you provided as an initialization and forced set the mean and std in the info.json file to be 0,1 respectively for all channels. It seems I can within one epoch immediately regenerate some of your results (in the sense that the segmentation on the training set is really good). However, currently, I only have access to CASIA-V4 dataset but not the rest of the dataset making it rather hard to properly understand the generalization performanec without having access to some other data never used in training. That unfortunately, might take a while given all the license paper work that should be done.

**Comment by DanielKerrigan** [comment]
It was only my bachelor's degree, so no Dr Daniel :)

I will try to find time outside of work to recreate that file.

I'm glad to hear that your results are looking better now. Unfortunately, I don't think we can give you the datasets ourselves.

**Comment by adelbibi** [comment]
No worries regarding the datasets. I'm in contact with the authors directly.
Please do let me know when you get to compute the mean/std json file. :-)

Best,

**Comment by DanielKerrigan** [comment]
Hi. Sorry for the wait. I have computed the mean and std over the training data:

```
{
    "mean": [
        0.6518711017992159,
        0.6518711017992159,
        0.6518711017992159
    ],
    "std": [
        0.22412408232719616,
        0.22412408232719616,
        0.22412408232719616
    ]
}
```

One thing to note is that the code in [`drn/segment.py`](https://github.com/fyu/drn/blob/96243c3068b47d2d78052d95ca11cf957f4ea02a/segment.py#L638-L639) divides the input images by 255 before doing the normalization using the mean and std. So the training and testing images should have values in the range [0, 255], but the values in `info.json` are expected to be in the range [0, 1].

I find this to be confusing and I wonder if I noticed this when I first trained and tested the models. I may have used the range [0, 255] for both the images and `info.json`.
